pydra/engine/specs.py

```diff
@@ -636,11 +636,12 @@ def _check_requires(self, fld, inputs):
 
         if "requires" in fld.metadata:
             # if requires is a list of list it is treated as el[0] OR el[1] OR...
-            if all([isinstance(el, list) for el in fld.metadata["requires"]]):
-                field_required_OR = fld.metadata["requires"]
+            required_fields = list(fld.metadata["requires"])
+            if all([isinstance(el, list) for el in required_fields]):
+                field_required_OR = required_fields
             # if requires is a list of tuples/strings - I'm creating a 1-el nested list
```

Comment:
I'm a little confused by this one. The type checking is divorced from the semantics. Why should this be a list of lists of tuples or strings? It's hard to reason about the effects of converting an arbitrary iterable to a list when the meaning of the tuples and strings (both iterables) are opaque.

Instead of having this here in this function, I'd rather have a separate function:

```Python
def canonicalize_requires(requires):
    """Canonicalize a requirement to a nested list of lists of requirements

    Fully specified requirements are a list of lists of strings or tuples that have the form
    ``[["abc", "def"], ...]`` or ``[[("abc", "def"), ("ghi", "jkl")], ...]``. For brevity, we permit the following
    forms.

    >>> # some example code

    Non-list iterables are also accepted:

    >>> canonicalize_requirements({"abc", "def"})
    [["abc", "def"]]
    """
    ...
```

Reply:
I did not want to change the semantics too much in this PR so I did not attempt to refactor the code too much. The code flow could be improved.

The idea remains that I should be able to do:

```python
metadata={
    ...,
    "requires": {"foo", "bar"},
}
```

or

```python
metadata={
    ...,
    "requires": ["foo", "bar"],
}
```

and the most straightfoward way was to normalize `Iterable` to `list` internally first.